libcudacxx/include/cuda/std/__bit/rotate.h

```diff
@@ -21,47 +21,61 @@
 #  pragma system_header
 #endif // no system header
 
-#include <cuda/std/__type_traits/enable_if.h>
+#include <cuda/std/__concepts/concept_macros.h>
+#include <cuda/std/__type_traits/is_constant_evaluated.h>
+#include <cuda/std/__type_traits/is_same.h>
 #include <cuda/std/__type_traits/is_unsigned_integer.h>
+#include <cuda/std/cstdint>
 #include <cuda/std/limits>
 
 _LIBCUDACXX_BEGIN_NAMESPACE_STD
 
-template <class _Tp>
-_CCCL_NODISCARD _LIBCUDACXX_HIDE_FROM_ABI constexpr _Tp __rotl(_Tp __t, unsigned int __cnt) noexcept
+template <typename _Tp>
+_CCCL_NODISCARD _LIBCUDACXX_HIDE_FROM_ABI constexpr _Tp __rotr(_Tp __t, int __cnt) noexcept
 {
-  static_assert(__cccl_is_unsigned_integer_v<_Tp>, "__rotl requires unsigned");
-  using __nlt = numeric_limits<_Tp>;
-
-  return ((__cnt % __nlt::digits) == 0)
-         ? __t
-         : (__t << (__cnt % __nlt::digits)) | (__t >> (__nlt::digits - (__cnt % __nlt::digits)));
+  constexpr auto __digits = numeric_limits<_Tp>::digits;
+  if (!_CUDA_VSTD::__cccl_default_is_constant_evaluated() && is_same_v<_Tp, uint32_t>)
+  {
+    NV_IF_TARGET(NV_IS_DEVICE, (return ::__funnelshift_r(__t, __t, __cnt);))
+  }
```

Comment:
We can make this slightly more efficient
```suggestion
  if constexpr(is_same_v<_Tp, uint32_t>) {
    if (!_CUDA_VSTD::__cccl_default_is_constant_evaluated())
    {
      NV_IF_TARGET(NV_IS_DEVICE, (return ::__funnelshift_r(__t, __t, __cnt);))
    }
  }
```

Applies below

Reply:
great suggestion. thanks